### ISSUE TITLE
Error: not a valid hash:undefined

### ISSUE THREAD

**Issue by christianbundy** [ssbc/patchbay#231]
I've been getting this sporadically for a few days, but I want to make sure it's documented. I'm not sure it's correlated with anything I'm doing in the UI, which makes me wonder whether it's happening in the background.

```
Error: not a valid hash:undefined
    at Object.<anonymous> (/home/christianbundy/Source/patchbay/node_modules/multiblob/index.js:156:19)
    at Object.hooked [as size] (/home/christianbundy/Source/patchbay/node_modules/hoox/index.js:10:15)
    at /home/christianbundy/Source/patchbay/node_modules/multiblob-http/index.js:58:13
    at handle (/home/christianbundy/Source/patchbay/node_modules/stack/stack.js:11:9)
    at /home/christianbundy/Source/patchbay/node_modules/stack/stack.js:13:11
    at /home/christianbundy/Source/patchbay/node_modules/ssb-ws/json-api.js:51:58
    at Object.want (/home/christianbundy/Source/patchbay/node_modules/ssb-blobs/inject.js:298:16)
    at Object.hooked [as want] (/home/christianbundy/Source/patchbay/node_modules/hoox/index.js:10:15)
    at /home/christianbundy/Source/patchbay/node_modules/ssb-ws/json-api.js:51:25
    at Object.has (/home/christianbundy/Source/patchbay/node_modules/ssb-blobs/inject.js:256:16)
```

**Comment by alanz** [comment]
Here is another one

```
Error: not a valid hash:undefined
    at Object.<anonymous> (/home/alanz/mysrc/ssb/patchbay/node_modules/multiblob/index.js:156:19)
    at Object.hooked [as size] (/home/alanz/mysrc/ssb/patchbay/node_modules/hoox/index.js:10:15)
    at /home/alanz/mysrc/ssb/patchbay/node_modules/multiblob-http/index.js:58:13
    at handle (/home/alanz/mysrc/ssb/patchbay/node_modules/stack/stack.js:11:9)
    at /home/alanz/mysrc/ssb/patchbay/node_modules/stack/stack.js:13:11
    at /home/alanz/mysrc/ssb/patchbay/node_modules/ssb-ws/json-api.js:51:58
    at Object.want (/home/alanz/mysrc/ssb/patchbay/node_modules/ssb-blobs/inject.js:298:16)
    at Object.hooked [as want] (/home/alanz/mysrc/ssb/patchbay/node_modules/hoox/index.js:10:15)
    at /home/alanz/mysrc/ssb/patchbay/node_modules/ssb-ws/json-api.js:51:25
    at Object.has (/home/alanz/mysrc/ssb/patchbay/node_modules/ssb-blobs/inject.js:256:16)
```

These only occur with current master, not with the `dark_crystal` branch, which is only a few commits behind, as I understand it.

**Comment by christianbundy** [comment]
It looks like these are coming from Electron requesting http://localhost:8989/blobs/get/undefined, which is a *great* opportunity to resolve a bug. I've made a small patch to make debugging much easier:

```diff
diff --git a/blob/sync/url.js b/blob/sync/url.js
index ea308df..fe24bbc 100644
--- a/blob/sync/url.js
+++ b/blob/sync/url.js
@@ -1,4 +1,5 @@
 const nest = require('depnest')
+const console = require('console')
 
 exports.gives = nest('blob.sync.url')
 
@@ -6,6 +7,11 @@ exports.create = function (api) {
   return nest('blob.sync.url', function (id) {
     // return id
 
-    return 'http://localhost:8989/blobs/get/' + id
+    if (id === undefined)  {
+      console.trace('value of "undefined" was passed as a blob ID')
+      return '#'
+    } else {
+      return 'http://localhost:8989/blobs/get/' + id
+    }
   })
 }
```

Here's the stacktrace I'm seeing:

```
Trace: value of "undefined" was passed as a blob ID
    at Object.<anonymous> (/home/christianbundy/Source/patchbay/blob/sync/url.js:11:15)
    at Object.url (/home/christianbundy/Source/patchbay/node_modules/depject/apply.js:18:30)
    at renderContent (/home/christianbundy/Source/patchbay/message/html/render/about.js:48:49)
    at Object.about (/home/christianbundy/Source/patchbay/message/html/render/about.js:29:16)
    at Object.render (/home/christianbundy/Source/patchbay/node_modules/depject/apply.js:18:30)
    at render (/home/christianbundy/Source/patchbay/app/page/public.js:67:33)
    at add (/home/christianbundy/Source/patchbay/node_modules/pull-scroll/index.js:42:15)
    at /home/christianbundy/Source/patchbay/node_modules/pull-scroll/index.js:71:11
    at /home/christianbundy/Source/patchbay/node_modules/pull-stream/sinks/drain.js:24:37
    at /home/christianbundy/Source/patchbay/node_modules/pull-stream/throughs/filter.js:17:11
    at /home/christianbundy/Source/patchbay/node_modules/pull-stream/throughs/filter.js:17:11
    at /home/christianbundy/Source/patchbay/node_modules/pull-stream/throughs/filter.js:17:11
    at /home/christianbundy/Source/patchbay/node_modules/pull-stream/throughs/filter.js:17:11
    at /home/christianbundy/Source/patchbay/node_modules/pull-stream/throughs/filter.js:17:11
    at /home/christianbundy/Source/patchbay/node_modules/pull-stream/throughs/filter.js:17:11
    at /home/christianbundy/Source/patchbay/node_modules/pull-abortable/index.js:46:11
```

I know enough to hide the symptoms, but I'd like to resolve the original problem. I'll keep working on it, but hopefully this foothold will help others who are debugging too. Good luck!

**Comment by alanz** [comment]
After applying that patch I got this on my console

```
STARTING app
(node:17484) [DEP0016] DeprecationWarning: 'GLOBAL' is deprecated, use 'global'
Trace: value of "undefined" was passed as a blob ID
    at Object.<anonymous> (/home/alanz/mysrc/ssb/patchbay/blob/sync/url.js:12:14)
    at Object.url (/home/alanz/mysrc/ssb/patchbay/node_modules/depject/apply.js:18:30)
    at renderContent (/home/alanz/mysrc/ssb/patchbay/message/html/render/about.js:48:49)
    at Object.about (/home/alanz/mysrc/ssb/patchbay/message/html/render/about.js:29:16)
    at Object.render (/home/alanz/mysrc/ssb/patchbay/node_modules/depject/apply.js:18:30)
    at render (/home/alanz/mysrc/ssb/patchbay/app/page/public.js:67:33)
    at add (/home/alanz/mysrc/ssb/patchbay/node_modules/pull-scroll/index.js:42:15)
    at HTMLDivElement.scroll (/home/alanz/mysrc/ssb/patchbay/node_modules/pull-scroll/index.js:50:7)
Error replicating with @AnnTkXjX2LkvuP3IRzd7EtGl2gwEx2lENy74pb1IE90=.ed25519:
   Error: parent stream is closing
    at PacketStream.stream (/home/alanz/mysrc/ssb/patchbay/node_modules/packet-stream/index.js:46:28)
    at Object.ws.remoteCall (/home/alanz/mysrc/ssb/patchbay/node_modules/muxrpc/stream.js:114:17)
    at /home/alanz/mysrc/ssb/patchbay/node_modules/muxrpc/index.js:45:15
    at remoteCall (/home/alanz/mysrc/ssb/patchbay/node_modules/muxrpc/api.js:26:19)
    at EventEmitter.obj.(anonymous function) [as createHistoryStream] (/home/alanz/mysrc/ssb/patchbay/node_modules/muxrpc/api.js:40:20)
    at replicate (/home/alanz/mysrc/ssb/patchbay/node_modules/scuttlebot/plugins/replicate/legacy.js:247:13)
    at EventEmitter.fallback (/home/alanz/mysrc/ssb/patchbay/node_modules/scuttlebot/plugins/replicate/legacy.js:307:7)
    at Object.onceWrapper (events.js:315:30)
    at emitOne (events.js:116:13)
    at EventEmitter.emit (events.js:211:7)
```

**Comment by christianbundy** [comment]
That's strange, mine isn't closing the stream. I've tracked the error down [here](https://github.com/ssbc/patchbay/blob/master/message/html/render/about.js#L48), although I'm not sure what to do. Usually ternary operations don't actually call the potential functions. For example:

```js
const slice2 = x => x.slice(2)
const a = null

a ? slice2(a) : false
// => false
```

On the other hand, `when()` doesn't look like it would keep `square(a)` from being called:

```js
when(a, slice2(a), false)
// => TypeError: null is not an object (evaluating 'x.slice')
```

I can think of two options off the top of my head:

1. Continue passing junk data to `url.js, but adding a runtime check for `undefined`. This adds a small bit of complexity but allows us to keep `about.js` as-is.
2. Patch `about.js` to replace `when` with a ternary function, which shouldn't (?) matter because `image` isn't an observable (?). I'd like to hear from others who are more familiar whether this is a good idea.

Anyone have any input? I'm not sure what the right call is here.

**Comment by alanz** [comment]
@christianbundy my javascript skills are close to zero. So I am not really in a position to help here.

**Comment by arj03** [comment]
Oh this is a very good find @christianbundy. I wonder how many booby traps are buried there!

**Comment by christianbundy** [comment]
@arj03 

I *think* only a few, hopefully! I've resolved all of the ones in this repo in #241, but I'm sure there are a handful of others. I wonder whether we're doing the same with other Mutant functions too...
